lib/file_system/backends/fs_poll.ex

```diff
@@ -0,0 +1,105 @@
+require Logger
+
+defmodule FileSystem.Backends.FSPoll do
+  @moduledoc """
+  FileSysetm backend for any OS, a GenServer that regularly scans file system to
+  detect changes and send them to the worker process.
+
+  ## Backend Options
+
+    * `:interval` (integer, default: 1000), polling interval
+
+  ## Use FSPoll Backend
+
+  Unlike other backends, polling backend is never automatically chosen in any
+  OS environment, despite being usable on all platforms.
+
+  To use polling backend, one has to explicitly specify in the backend option.
+  """
+
+  use GenServer
+  @behaviour FileSystem.Backend
+
+  def bootstrap, do: :ok
+
+  def supported_systems do
+    [{:unix, :linux}, {:unix, :freebsd}, {:unix, :darwin}, {:win32, :nt}]
+  end
+
+  def known_events do
+    [:created, :deleted, :modified]
+  end
+
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, [])
+  end
+
+  def init(args) do
+    {worker_pid, rest} = Keyword.pop(args, :worker_pid)
+    {interval, rest} = Keyword.pop(rest, :interval, 1000)
+    {dirs, _unknown_opts} = Keyword.pop(rest, :dirs)
+
+    Logger.info("Polling file changes every #{interval}ms...")
+    send(self(), :first_check)
+
+    {:ok, {worker_pid, dirs, interval, %{}}}
+  end
+
+  def handle_info(:first_check, {worker_pid, dirs, interval, _empty_map}) do
+    schedule_check(interval)
+    {:noreply, {worker_pid, dirs, interval, files_mtimes(dirs)}}
+  end
+
+  def handle_info(:check, {worker_pid, dirs, interval, stale_mtimes}) do
+    fresh_mtimes = files_mtimes(dirs)
+
+    diff(stale_mtimes, fresh_mtimes)
+    |> Tuple.to_list
+    |> Enum.zip([:created, :deleted, :modified])
+    |> Enum.each(&report_change(&1, worker_pid))
+
+    schedule_check(interval)
+    {:noreply, {worker_pid, dirs, interval, fresh_mtimes}}
+  end
+
+  defp schedule_check(interval) do
+    Process.send_after(self(), :check, interval)
+  end
+
+  defp files_mtimes(dirs, files_mtimes_map \\ %{}) do
+    Enum.reduce(dirs, files_mtimes_map, fn dir, map ->
+      case File.stat!(dir) do
+        %{type: :regular, mtime: mtime} ->
+          Map.put(map, dir, mtime)
+        %{type: :directory} ->
+          dir
+          |> Path.join("**/*")
+          |> Path.wildcard
+          |> files_mtimes(map)
```

Comment:
`Path.wildcard("#{base}/**/*")` will process sub dirs, and `files_mtimes` will do `wildcard` again, so file maybe processed multiple times.

Reply:
Right!